src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,392 @@
+        /// </para>
+        /// </summary>
+        public async Task InitializeAsync()
+        {
+            this.logger.LogTrace("()");
+
+            if (this.storeSettings.ReIndex)
+            {
+                this.logger.LogTrace("(-)[REINDEX_NOT_SUPPORTED]");
+                throw new NotSupportedException("Re-indexing the block store in currently not supported.");
+            }
+
+            ChainedHeader initializationTip = this.chain.GetBlock(this.blockRepository.BlockHash);
+            this.SetStoreTip(initializationTip);
+            
+            if (this.storeTip == null)
+                await this.RecoverStoreTipAsync().ConfigureAwait(false);
+
+            this.logger.LogDebug("Initialized block store tip at '{0}'.", this.storeTip);
+
+            if (this.storeSettings.TxIndex != this.blockRepository.TxIndex)
+            {
+                if (this.storeTip != this.chain.Genesis)
+                {
+                    this.logger.LogTrace("(-)[REBUILD_REQUIRED]");
+                    throw new BlockStoreException("You need to rebuild the block store database using -reindex-chainstate to change -txindex");
+                }
+
+                if (this.storeSettings.TxIndex)
+                    await this.blockRepository.SetTxIndexAsync(this.storeSettings.TxIndex).ConfigureAwait(false);
+            }
+            
+            // Throw if block store was initialized after the consensus.
+            // This is needed in order to rewind consensus in case it is initialized ahead of the block store.
+            if (this.chainState.ConsensusTip != null)
+            {
+                this.logger.LogCritical("Block store initialized after the consensus!");
+                this.logger.LogTrace("(-)[INITIALIZATION_ERROR]");
+                throw new BlockStoreException("Block store initialized after consensus!");
+            }
+
+            // Start dequeuing.
+            this.currentBatchSizeBytes = 0;
+            this.dequeueLoopTask = this.DequeueBlocksContinuouslyAsync();
+
+            this.logger.LogTrace("(-)");
+        }
+        
+        /// <summary>Sets the internal store tip and exposes the store tip to other components through the chain state.</summary>
+        private void SetStoreTip(ChainedHeader newTip)
+        {
+            this.storeTip = newTip;
+            this.chainState.BlockStoreTip = newTip;
+        }
+
+        /// <summary>
+        /// Sets block store tip to the last block that exists both in the repository and in the <see cref="ConcurrentChain"/>.
+        /// </summary>
+        private async Task RecoverStoreTipAsync()
+        {
+            this.logger.LogTrace("()");
+
+            var blockStoreResetList = new List<uint256>();
+            uint256 resetBlockHash = this.blockRepository.BlockHash;
+            Block resetBlock = await this.blockRepository.GetAsync(resetBlockHash).ConfigureAwait(false);
+
+            while (this.chain.GetBlock(resetBlockHash) == null)
```

Comment:
Thinking about it perhaps we should limit the amount of reorg going back so the node does not reorg 100k blocks by mistake?

The problem is what if the block store crashed with a big batch of blocks, so the limit should be set to the max block batch size.

Reply:
We don't need a limit here. 
It's not possible to reorg more than 1 minute worth of downloaded blocks. 

If we set a limit- we might not reorg enough and this will case problems.

Reply:
yeah but if a user accidentally deletes the store folder they will rewind consensus.

Reply:
Ah actually not, as the StoreHeader will not be set in ChainState

Reply:
>yeah but if a user accidentally deletes the store folder they will rewind consensus.

What should happen so the user accidentally deletes the part of the database? 

When you remove the registry from windows installation you don't expect it to recover on it's own 🙂